crates/chain-state/src/memory_overlay.rs

```diff
@@ -7,14 +7,26 @@ use alloy_primitives::{
 use reth_errors::ProviderResult;
 use reth_primitives::{Account, Bytecode};
 use reth_storage_api::{
-    AccountReader, BlockHashReader, StateProofProvider, StateProvider, StateProviderBox,
-    StateRootProvider, StorageRootProvider,
+    AccountReader, BlockHashReader, StateProofProvider, StateProvider, StateRootProvider,
+    StorageRootProvider,
 };
 use reth_trie::{
     updates::TrieUpdates, AccountProof, HashedPostState, HashedStorage, MultiProof, TrieInput,
 };
 use std::sync::OnceLock;
 
+/// A state provider that stores references to in-memory blocks along with their state as well as a
+/// reference of the historical state provider for fallback lookups.
+#[allow(missing_debug_implementations)]
+pub struct MemoryOverlayStateProviderRef<'a> {
+    /// Historical state provider for state lookups that are not found in in-memory blocks.
+    pub(crate) historical: Box<dyn StateProvider + 'a>,
+    /// The collection of executed parent blocks. Expected order is newest to oldest.
+    pub(crate) in_memory: Vec<ExecutedBlock>,
+    /// Lazy-loaded in-memory trie data.
+    pub(crate) trie_state: OnceLock<MemoryOverlayTrieState>,
+}
```

Comment:
I think we don't necessarily have to duplicate this.

we can make the `MemoryOverlayStateProvider`have a field of `MemoryOverlayStateProviderRef<'static>` then we don't need the macro I believe

but we can do this separately